### PR TITLE
Add allowed method for Lint/Debugger

### PR DESCRIPTION
**Pull request by mfbmina**
This PR fixes https://github.com/rubocop-hq/rubocop/issues/8927.

It allows the user to add AllowedMethods for Lint/Debugger cop.

- [ ] Wrote good commit messages.
- [ ]  Commit message starts with [Fix #issue-number] (if the related issue exists).
- [ ]  Feature branch is up-to-date with master (if not - rebase it).
- [ ]  Squashed related commits together.
- [ ]  Added tests.
- [ ]  Added an entry to the Changelog if the new code introduces user-observable changes. See changelog entry format.
- [ ]  The PR relates to only one subject with a clear title and description in grammatically correct, complete sentences.
- [ ]  Run bundle exec rake default. It executes all tests and RuboCop for itself, and generates the documentation.

### CODE REVIEW

**Comment by bbatsov** [comment]
You should also update the docs and the default configuration. And remove the hardcoded methods.

**Comment by mfbmina** [comment]
@bbatsov please check again!

File: config/default.yml [hunk 1]
```diff
@@ -1415,6 +1415,18 @@ Lint/Debugger:
   Enabled: true
   VersionAdded: '0.14'
   VersionChanged: '0.49'
+  AllowedMethods:
```
Comment (robotdana) [inline config/default.yml:1418]: this seems the opposite of Allowed, should this be `DisallowedMethods`? `ReportedMethods`? `DebuggerMethods`? just `Methods`?

Comment (bbatsov) [reply config/default.yml:1418]: Yep, you're toally right. I like the name `DebuggerMethods` best, although I'm also fine with "ForbiddenMethods".

Comment (mfbmina) [reply config/default.yml:1418]: I also felt like this when implementing... shall we go with `DebuggerMethods` then?

Comment (bbatsov) [reply config/default.yml:1418]: Yep.

**Comment by bbatsov** [comment]
> please check again!

Yout still have replace the hardcoded node patterns with something that's generated dynamically based on the cop's configuration.

**Comment by mfbmina** [comment]
> Yout still have replace the hardcoded node patterns with something that's generated dynamically based on the cop's configuration.

My solution for that was adding a `ForbiddenReceivers` options which will help to specify which receivers are not allowed, like `binding` or `Kernel`. 

The only edge case that I couldn't cover and had to add a special check when was `Kernel.save_and_open_page` which should be allowed even tho `Kernel` is not an allowed receiver and `save_and_open_page` is a debugger method.

**Comment by mfbmina** [comment]
@bbatsov @robotdana what you guys think?

File: config/default.yml [hunk 2]
```diff
@@ -1415,6 +1415,22 @@ Lint/Debugger:
   Enabled: true
   VersionAdded: '0.14'
   VersionChanged: '0.49'
+  ForbiddenReceivers:
```
Comment (bbatsov) [inline config/default.yml:1418]: We can name this "DebuggerReceivers" for consistency.

File: lib/rubocop/cop/mixin/forbidden_receivers.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # This module encapsulates the ability to forbid certain method receivers
+    # when parsing.
+    module ForbiddenReceivers
```
Comment (bbatsov) [inline lib/rubocop/cop/mixin/forbidden_receivers.rb:7]: I don't think we need a module for this, so I'd fold it back into the cop.

**Comment by bbatsov** [comment]
I think we're pretty close to merging this. I left a couple of small remarks. You'll also need to update the cop description to mention the configuration options and the rebase on top of master.

File: lib/rubocop.rb [hunk 4]
```diff
@@ -69,6 +69,7 @@
 require_relative 'rubocop/cop/mixin/configurable_formatting'
 require_relative 'rubocop/cop/mixin/configurable_naming'
 require_relative 'rubocop/cop/mixin/configurable_numbering'
+require_relative 'rubocop/cop/mixin/debugger_methods'
```
Comment (bbatsov) [inline lib/rubocop.rb:72]: I think this mixin should be inlined as well. I doubt we'll ever reuse it.

File: lib/rubocop/cop/lint/debugger.rb [hunk 5]
```diff
@@ -72,8 +52,26 @@ def message(node)
           format(MSG, source: node.source)
         end
 
-        def binding_irb?(node)
-          binding_irb_call?(node)
+        def with_receiver?(node)
+          !!node.receiver
+        end
+
+        def special_rule?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/debugger.rb:59]: I don't get what's the purpose of this `special_rule?` method.

File: lib/rubocop/cop/lint/debugger.rb [hunk 6]
```diff
@@ -72,8 +52,26 @@ def message(node)
...
+          return false unless node.receiver.is_a?(RuboCop::AST::ConstNode)
+          return true if node.receiver.const_name == 'Kernel' && node.method?(:save_and_open_page)
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/debugger.rb:61]: In particular it's very weird we hardcoded this.

File: lib/rubocop/cop/lint/debugger.rb [hunk 7]
```diff
@@ -33,35 +34,14 @@ module Lint
       #     do_something
       #   end
       class Debugger < Base
-        MSG = 'Remove debugger entry point `%<source>s`.'
-
-        RESTRICT_ON_SEND = %i[
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/debugger.rb:38]: You should keep this constant, as it's special for RuboCop (it limits the `on_send` callbacks). Just initialize it with the data from the config.

Comment (mfbmina) [reply lib/rubocop/cop/lint/debugger.rb:38]: The problem if I hardcode the values from the configuration in this constant is not recognizing new values from the config.

This test, for example, start failing: https://github.com/rubocop-hq/rubocop/blob/12bd06882eaca12bd33c77f31e8d976bc3ad0bed/spec/rubocop/cop/lint/debugger_spec.rb#L15-L20

What I did, for now, is initializing the constant but with an empty array.

File: lib/rubocop/cop/lint/debugger.rb (hunk 5 above)
Comment (mfbmina) [reply lib/rubocop/cop/lint/debugger.rb:59]: this is an edge case that I couldn't find a better fix. Maybe you can help me:

Even tho `Kernel` is a receiver and `save_and_open_page` is a debugger method together they should not be an offense.

Here is the test case for it: https://github.com/rubocop-hq/rubocop/blob/12bd06882eaca12bd33c77f31e8d976bc3ad0bed/spec/rubocop/cop/lint/debugger_spec.rb#L164-L166

File: lib/rubocop/cop/lint/debugger.rb (hunk 6 above)
Comment (mfbmina) [reply lib/rubocop/cop/lint/debugger.rb:61]: I also don't like it. Look above why I did it.

File: lib/rubocop/cop/lint/debugger.rb (hunk 7 above)
Comment (bbatsov) [reply lib/rubocop/cop/lint/debugger.rb:38]: Ah, yeah - that's a fair point.

File: lib/rubocop/cop/lint/debugger.rb (hunk 5 above)
Comment (bbatsov) [reply lib/rubocop/cop/lint/debugger.rb:59]: I'd just drop this test case. For such a small thing people should add a local exclusion rule. It just doesn't fit with having with our goal to have a configurable cop.

Comment (mfbmina) [reply lib/rubocop/cop/lint/debugger.rb:59]: Cool! I will drop it! It was introduced here. https://github.com/rubocop-hq/rubocop/commit/83902a6a4b957130abcabdebe1c266a12ec5d138

File: lib/rubocop/cop/lint/debugger.rb [hunk 8]
```diff
@@ -72,8 +51,23 @@ def message(node)
           format(MSG, source: node.source)
         end
 
-        def binding_irb?(node)
-          binding_irb_call?(node)
+        def with_receiver?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/debugger.rb:54]: I see you're using this method only once, so I'd inline it in `forbidden_receiver?`.

File: lib/rubocop/cop/lint/debugger.rb [hunk 9]
```diff
@@ -72,8 +51,23 @@ def message(node)
...
+          !!node.receiver
+        end
+
+        def debugger_method?(name)
+          cop_config.fetch('DebuggerMethods', []).include?(name.to_s)
+        end
+
+        def forbidden_receiver?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/debugger.rb:62]: I'd name this `debugger_receiver?`.

**Review by bbatsov — APPROVED** [review]
